packages/nodes-base/utils/connection-pool-manager.ts

```diff
@@ -0,0 +1,135 @@
+import { createHash } from 'crypto';
+
+let instance: ConnectionPoolManager;
+
+// 5 minutes
+const ttl = 5 * 60 * 1000;
+
+// 1 minute
+const cleanUpInterval = 60 * 1000;
+
+type Options = {
+	credentials: unknown;
+	nodeType: string;
+	nodeVersion?: string;
+};
+
+export class ConnectionPoolManager {
+	/**
+	 * Gets the singleton instance of the ConnectionPoolManager.
+	 * Creates a new instance if one doesn't exist.
+	 */
+	static getInstance(): ConnectionPoolManager {
+		if (instance) {
+			return instance;
+		}
+
+		instance = new ConnectionPoolManager();
+
+		return instance;
+	}
+
+	/**
+	 * Private constructor that initializes the connection pool manager.
+	 * Sets up cleanup handlers for process exit and stale connections.
+	 */
+	private constructor(
+		private map: Map<
+			string,
+			{ pool: unknown; cleanUpHandler: (pool: unknown) => Promise<void>; lastUsed: number }
+		> = new Map(),
+	) {
+		// Close all SSH connections when the process exits
+		process.on('exit', () => this.onShutdown());
+
+		// Regularly close stale SSH connections
+		setInterval(() => this.cleanupStaleConnections(), cleanUpInterval);
+	}
```

Comment:
having this in the constructor is a bit confusing. can we move this bit out?
```suggestion
	private map: Map<
			string,
			{ pool: unknown; cleanUpHandler: (pool: unknown) => Promise<void>; lastUsed: number }
		> = new Map()

	/**
	 * Private constructor that initializes the connection pool manager.
	 * Sets up cleanup handlers for process exit and stale connections.
	 */
	private constructor() {
		// Close all SSH connections when the process exits
		process.on('exit', () => this.onShutdown());

		// Regularly close stale SSH connections
		setInterval(() => this.cleanupStaleConnections(), cleanUpInterval);
	}
```

also, we could extract the `{ pool: unknown; cleanUpHandler: (pool: unknown) => Promise<void>; lastUsed: number }` bit into a type, and add jsdocs for each property (for future readability).